MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/query/FingerDrawQueryActivity.java

```diff
@@ -0,0 +1,391 @@
+  private static final String LINE_COLOR = "#a0861c";
+  private static final float LINE_WIDTH = 5f;
+  private static final float LINE_OPACITY = 1f;
+  private static final float FILL_OPACITY = .4f;
+
+  private MapView mapView;
+  private MapboxMap mapboxMap;
+  private FeatureCollection searchPointFeatureCollection;
+  private GeoJsonSource drawLineSource;
+  private GeoJsonSource fillPolygonSource;
+  private List<Point> freehandTouchPointListForPolygon = new ArrayList<>();
+  private List<Point> freehandTouchPointListForLine = new ArrayList<>();
+  private List<List<Point>> polygonList;
+  private Point firstScreenTouchPoint;
+  private boolean showSearchDataLocations = true;
+  private boolean drawSingleLineOnly = false;
+
+  private View.OnTouchListener customOnTouchListener = new View.OnTouchListener() {
+    @Override
+    public boolean onTouch(View view, MotionEvent motionEvent) {
+
+      LatLng latLngTouchCoordinate = mapboxMap.getProjection().fromScreenLocation(
+        new PointF(motionEvent.getX(), motionEvent.getY()));
+
+      Point screenTouchPoint = Point.fromLngLat(latLngTouchCoordinate.getLongitude(),
+        latLngTouchCoordinate.getLatitude());
+
+      // Draw the line on the map as the finger is dragged along the map
+      freehandTouchPointListForLine.add(screenTouchPoint);
+      drawLineSource = mapboxMap.getStyle().getSourceAs(FREEHAND_DRAW_LINE_LAYER_SOURCE_ID);
+      drawLineSource.setGeoJson(LineString.fromLngLats(freehandTouchPointListForLine));
+
+      // Draw a polygon area if drawSingleLineOnly == false
+      if (!drawSingleLineOnly) {
+        if (freehandTouchPointListForPolygon.size() == 0) {
+          firstScreenTouchPoint = screenTouchPoint;
+        }
+        if (freehandTouchPointListForPolygon.size() < 3) {
+          freehandTouchPointListForPolygon.add(screenTouchPoint);
+        } else if (freehandTouchPointListForPolygon.size() == 3) {
+          freehandTouchPointListForPolygon.add(screenTouchPoint);
+          freehandTouchPointListForPolygon.add(firstScreenTouchPoint);
+        } else {
+          freehandTouchPointListForPolygon.remove(freehandTouchPointListForPolygon.size() - 1);
+          freehandTouchPointListForPolygon.add(screenTouchPoint);
+          freehandTouchPointListForPolygon.add(firstScreenTouchPoint);
+        }
+
+        // Create and show a FillLayer polygon where the search area is
+        fillPolygonSource = mapboxMap.getStyle().getSourceAs(FREEHAND_DRAW_FILL_LAYER_SOURCE_ID);
+        polygonList = new ArrayList<>();
+        polygonList.add(freehandTouchPointListForPolygon);
+        fillPolygonSource.setGeoJson(Polygon.fromLngLats(polygonList));
+      }
+
+      // Take certain actions when the drawing is done
+      if (motionEvent.getAction() == MotionEvent.ACTION_UP) {
+
+        // If drawing polygon, add the first screen touch point to the end of
+        // the LineLayer list so that it's
+        if (!drawSingleLineOnly) {
```

Comment:
Not needed case.

Reply:
Needed because the example offers a button to toggle between drawing just a line or drawing a line+polygon.

Reply:
The line geometry doesn't have to be closed.

Reply:
Also, the source doesn't seem to be updated after this addition anyway.

Reply:
>Needed because the example offers a button to toggle between drawing just a line or drawing a line+polygon.

Made ⬆️  comment because I was looking at the wrong `if (!drawSingleLineOnly) {` statement.

Regarding the one you're talking about, I close the line geometry because if I don't, the end result of drawing a polygon looks like 👇 

![device-2019-05-30-105345](https://user-images.githubusercontent.com/4394910/58652808-3f16b180-82c9-11e9-960c-f7e249374d9e.png)

I'd rather close the line geometry to have it follow all edges of the polygon (see below), rather than the line geometry being "open".

![device-2019-05-30-105942](https://user-images.githubusercontent.com/4394910/58653218-1511bf00-82ca-11e9-8711-2aa758a39306.png)

Reply:
Ah alright, this makes perfect sense.